### PR TITLE
Persist skirmish settings between sessions

### LINKED ISSUE #4122: Remember lobby option checkbox states
I think it could be interesting if Map Options (shroud, fog of war, crates appear...) could be remembered after modificating them.

This way, a player could create new games according to his preferences (and without always having to re-enable them each time a new game is created).

Same for the Debug (Developer Mode), checkboxes could be saved too.

### LINKED ISSUE #4787: Remember faction for the mod?
Would be nice, but simple addition.

So if i play almost always with Allies, i dont have to select it every time i play.

### PR DESCRIPTION
**Pull request by pchote**
Closes #4122
Closes #4787

### CODE REVIEW

**Comment by pchote** [comment]
Rewritten to split the skirmish logic from LobbyCommands into its own ServerTrait.

File: OpenRA.Game/Server/Server.cs [hunk 1]
```diff
@@ -40,8 +40,9 @@ public enum ServerState
 	public enum ServerType
 	{
 		Local = 0,
-		Multiplayer = 1,
-		Dedicated = 2
+		Skirmish = 1,
+		Multiplayer = 2,
+		Dedicated = 3
```
Comment (PunkPun) [inline OpenRA.Game/Server/Server.cs:45]: then rename `Local` to `Mission` ?

Comment (pchote) [reply OpenRA.Game/Server/Server.cs:45]: `Local` is used for more than just missions (e.g. the map editor "Play Map" option).

File: OpenRA.Mods.Common/ServerTraits/SkirmishLogic.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+using OpenRA.Primitives;
+using OpenRA.Server;
+using OpenRA.Traits;
+using S = OpenRA.Server.Server;
+
+namespace OpenRA.Mods.Common.Server
+{
+	public class SkirmishLogic : ServerTrait, IClientJoined, INotifySyncLobbyInfo
+	{
+		class SkirmishSlot
```
Comment (PunkPun) [inline OpenRA.Mods.Common/ServerTraits/SkirmishLogic.cs:25]:
```suggestion
		readonly struct SkirmishSlot
```

Why not struct?

File: OpenRA.Game/Server/Server.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Game/Server/Server.cs:45]: perhaps then split it into Mission and Editor. We will want a save system for missions as well

Comment (pchote) [reply OpenRA.Game/Server/Server.cs:45]: That split can be done in the PR that adds a save system for missions.

**Review by PunkPun — COMMENTED** [review]
Once there's a save I cannot join skirmish. 

I also got this crash when changing slot
```
Exception of type `System.ArgumentNullException`: Value cannot be null. (Parameter 'key')
   at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at OpenRA.Mods.Common.Widgets.Logic.LobbyUtils.SetupEditableFactionWidget(Widget parent, Slot s, Client c, OrderManager orderManager, Dictionary`2 factions) in OpenRA/OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:line 554
   at OpenRA.Mods.Common.Widgets.Logic.LobbyLogic.UpdatePlayerList() in OpenRA/OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:line 722
   at OpenRA.Network.UnitOrders.ProcessOrder(OrderManager orderManager, World world, Int32 clientId, Order order) in OpenRA/OpenRA.Game/Network/UnitOrders.cs:line 316
   at OpenRA.Network.OrderManager.ReceiveImmediateOrders(Int32 clientId, OrderPacket orders) in OpenRA/OpenRA.Game/Network/OrderManager.cs:line 180```

**Review by PunkPun — COMMENTED** [review]
IsSkirmish should be true when in lobby from the map editors PlayMap option

**Comment by pchote** [comment]
> IsSkirmish should be true when in lobby from the map editors PlayMap option

No, as I mentioned above, this must be false to prevent the server logic from overriding the map being edited, and for the other state set in the editor from overriding the skirmish data.

**Comment by pchote** [comment]
> Once there's a save I cannot join skirmish.
> I also got this crash when changing slot

Can you please post the skirmish yaml? I'm not able to reproduce this locally.

File: OpenRA.Mods.Common/ServerTraits/SkirmishLogic.cs (hunk 2 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/ServerTraits/SkirmishLogic.cs:25]: oh, field loader doesn't work with structs? That should be documented. Saving structs works just fine

That's why I was getting https://github.com/OpenRA/OpenRA/pull/21206#pullrequestreview-1726306260

**Comment by RoosterDragon** [comment]
Overall seems good.

If I modify the saved file to create an invalid option (e.g. set `gamespeed: no-such-option`), then I can no longer open the Skirmish menu without an error
image

I don't think we need to protect against users messing with the file as such, but in the future if we change the lobby options or anything else that gets saved then the file on disk might not be valid any more. I think we need to recover from this scenario. Whether we drop the bad file entirly or try and recover individual options in a piecemeal fashion I don't mind.

**Comment by pchote** [comment]
Rebased and fixed.

**Review by PunkPun — APPROVED** [review]
Overall LGTM

**Comment by PunkPun** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a47e0aad6707e71746cd09183c073e97831970d0...e8d200f76b51ec1a3ad56d33724942e4be56a1cc)

**Pull request merged**